src/kafka/connection/kafka_connection.hh

```diff
@@ -0,0 +1,137 @@
+    int32_t _correlation_id;
+    api_versions_response _api_versions;
+
+    template<typename RequestType>
+    future<int32_t> send_request(const RequestType &request, int16_t api_version) {
+        auto assigned_correlation_id = _correlation_id++;
+
+        std::vector<char> header;
+        boost::iostreams::back_insert_device<std::vector<char>> header_sink{header};
+        boost::iostreams::stream<boost::iostreams::back_insert_device<std::vector<char>>> header_stream{header_sink};
+        request_header req_header;
+        req_header._api_key = RequestType::API_KEY;
+        req_header._api_version = api_version;
+        req_header._correlation_id = assigned_correlation_id;
+        req_header._client_id = _client_id;
+        req_header.serialize(header_stream, 0);
+        header_stream.flush();
+
+        std::vector<char> payload;
+        boost::iostreams::back_insert_device<std::vector<char>> payload_sink{payload};
+        boost::iostreams::stream<boost::iostreams::back_insert_device<std::vector<char>>> payload_stream{payload_sink};
+        request.serialize(payload_stream, api_version);
+        payload_stream.flush();
+
+        std::vector<char> message;
+        boost::iostreams::back_insert_device<std::vector<char>> message_sink{message};
+        boost::iostreams::stream<boost::iostreams::back_insert_device<std::vector<char>>> message_stream{message_sink};
+
+        kafka_int32_t message_size(header.size() + payload.size());
+        message_size.serialize(message_stream, 0);
+        message_stream.write(header.data(), header.size());
+        message_stream.write(payload.data(), payload.size());
+        message_stream.flush();
+
+        temporary_buffer<char> message_buffer{message.data(), message.size()};
+        return _connection->write(message_buffer.clone())
+                .then([assigned_correlation_id] { return assigned_correlation_id; });
+    }
+
+    template<typename RequestType>
+    future<typename RequestType::response_type> receive_response(int32_t correlation_id, int16_t api_version) {
+        return _connection->read(4).then([] (temporary_buffer<char> response_size) {
+            boost::iostreams::stream<boost::iostreams::array_source> response_size_stream
+                    (response_size.get(), response_size.size());
+
+            kafka_int32_t size;
+            size.deserialize(response_size_stream, 0);
+            return *size;
+        }).then([this] (int32_t response_size) {
+            return _connection->read(response_size);
+        }).then([correlation_id, api_version] (temporary_buffer<char> response) {
+            boost::iostreams::stream<boost::iostreams::array_source> response_stream
+                    (response.get(), response.size());
+
+            kafka::response_header response_header;
+            response_header.deserialize(response_stream, 0);
+            if (*response_header._correlation_id != correlation_id) {
+                throw parsing_exception();
```

Comment:
When can this occur?

Reply:
In the future we may send many messages concurrently on the same TCP socket. This is just a sanity check to see if messages are sent/received in the right order.